object-api.go

```diff
@@ -0,0 +1,322 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/minio/minio/pkg/mimedb"
+	"github.com/minio/minio/pkg/probe"
+	"github.com/minio/minio/pkg/safe"
+)
+
+type objectAPI struct {
+	storage StorageAPI
+}
+
+func newObjectLayer(storage StorageAPI) *objectAPI {
+	return &objectAPI{storage}
+}
+
+/// Bucket operations
+
+// MakeBucket - make a bucket.
+func (o objectAPI) MakeBucket(bucket string) *probe.Error {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if e := o.storage.MakeVol(bucket); e != nil {
+		if e == errVolumeExists {
+			return probe.NewError(BucketExists{Bucket: bucket})
+		}
+		return probe.NewError(e)
+	}
+	return nil
+}
+
+// GetBucketInfo - get bucket info.
+func (o objectAPI) GetBucketInfo(bucket string) (BucketInfo, *probe.Error) {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return BucketInfo{}, probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	vi, e := o.storage.StatVol(bucket)
+	if e != nil {
+		if e == errVolumeNotFound {
+			return BucketInfo{}, probe.NewError(BucketNotFound{Bucket: bucket})
+		}
+		return BucketInfo{}, probe.NewError(e)
+	}
+	return BucketInfo{
+		Name:    vi.Name,
+		Created: vi.Created,
+	}, nil
+}
+
+// ListBuckets - list buckets.
+func (o objectAPI) ListBuckets() ([]BucketInfo, *probe.Error) {
+	var bucketInfos []BucketInfo
+	vols, e := o.storage.ListVols()
+	if e != nil {
+		return nil, probe.NewError(e)
+	}
+	for _, vol := range vols {
+		if !IsValidBucketName(vol.Name) {
+			continue
+		}
+		bucketInfos = append(bucketInfos, BucketInfo{vol.Name, vol.Created})
+	}
+	return bucketInfos, nil
+}
+
+// DeleteBucket - delete a bucket.
+func (o objectAPI) DeleteBucket(bucket string) *probe.Error {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	if e := o.storage.DeleteVol(bucket); e != nil {
+		if e == errVolumeNotFound {
+			return probe.NewError(BucketNotFound{Bucket: bucket})
+		}
+		return probe.NewError(e)
+	}
+	return nil
+}
+
+/// Object Operations
+
+// GetObject - get an object.
+func (o objectAPI) GetObject(bucket, object string, startOffset int64) (io.ReadCloser, *probe.Error) {
+	// Verify if bucket is valid.
+	if !IsValidBucketName(bucket) {
+		return nil, probe.NewError(BucketNameInvalid{Bucket: bucket})
+	}
+	// Verify if object is valid.
+	if !IsValidObjectName(object) {
+		return nil, probe.NewError(ObjectNameInvalid{Bucket: bucket, Object: object})
+	}
+	r, e := o.storage.ReadFile(bucket, object, startOffset)
```

Comment:
> As we pass bucket and object as is to storage, does it mean any storage should aware of such input arguments bucket/object names?
> 
> For an example, if object="path/to/my/object" works fine in windows and linux as file path. for object="path\to\my\object" windows has file path whereas in linux it's a file
> 
> I also see if a function in windows returning path uses \ eg path\to\my\object
> There is no object which has backward slash, while s3 supports it we don't

Adding to this there was a bug we discussed long time back - https://github.com/minio/minio/issues/1183 

Which we haven't addressed yet.